controllers/druid/supervisorspec_controller.go

```diff
@@ -0,0 +1,608 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	druidv1alpha1 "github.com/datainfrahq/druid-operator/apis/druid/v1alpha1"
+)
+
+type DruidApiSupervisor struct {
+	Id            string          `json:"id"`
+	State         string          `json:"state"`
+	DetailedState string          `json:"detailedState"`
+	Healthy       bool            `json:"healthy"`
+	Spec          json.RawMessage `json:"spec"`
+	Suspended     bool            `json:"suspended"`
+}
+
+type DruidApiSupervisorSpecMinDataSchema struct {
+	DataSource string `json:"dataSource"`
+}
+
+type DruidApiSupervisorSpecMin struct {
+	DataSchema DruidApiSupervisorSpecMinDataSchema `json:"dataSchema"`
+}
+
+type SupervisorSpecStateEntry struct {
+	Id                   string `json:"id"`
+	DruidServiceEndpoint string `json:"druidHost"`
+}
+
+const (
+	ActionCreate                  = "create"
+	ActionUpdate                  = "update"
+	ActionDelete                  = "delete"
+	SupervisorSpecConfigMap       = "supervisor-specs-controller"
+	SupervisorListUrlPattern      = "http://%s:8088/druid/indexer/v1/supervisor"
+	SupervisorUrlPattern          = "http://%s:8088/druid/indexer/v1/supervisor/%s"
+	SupervisorTerminateUrlPattern = "http://%s:8088/druid/indexer/v1/supervisor/%s/terminate"
+	SupervisorResumeUrlPattern    = "http://%s:8088/druid/indexer/v1/supervisor/%s/resume"
+	SupervisorSuspendUrlPattern   = "http://%s:8088/druid/indexer/v1/supervisor/%s/suspend"
+)
+
+// SupervisorSpecReconciler reconciles a SupervisorSpec object
+type SupervisorSpecReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=druid.apache.org,resources=supervisorspecs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=druid.apache.org,resources=supervisorspecs/status,verbs=get;update;patch
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SupervisorSpecReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&druidv1alpha1.SupervisorSpec{}).
+		Complete(r)
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the SupervisorSpec object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *SupervisorSpecReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithName(string(uuid.NewUUID()))
+	_ = log.WithValues("supervisorspec", req.NamespacedName)
```

Comment:
can be removed?

Reply:
why? It (log) is setting the value on itself inside WithValues via `l.setSink(l.sink.WithValues(keysAndValues...))`.

Reply:
log.WithValues() return a new logger

Reply:
You're right, i did see a pointer where there wasn't any.